airflow-core/src/airflow/task/pessimistic_task_selector.py

```diff
@@ -0,0 +1,204 @@
+            .label("total_tis_per_dagrun_count")
+        )
+        tis_per_dag_count = (
+            func.row_number()
+            .over(partition_by=(TI.dag_id, TI.task_id), order_by=priority_order)
+            .label("tis_per_dag_count")
+        )
+        mapped_tis_per_task_run_count = (
+            func.row_number()
+            .over(partition_by=(TI.dag_id, TI.run_id, TI.task_id), order_by=priority_order)
+            .label("mapped_tis_per_dagrun_count")
+        )
+        pool_slots_taken = (
+            func.sum(TI.pool_slots)
+            .over(partition_by=(TI.pool), order_by=priority_order)
+            .label("pool_slots_taken_sum")
+        )
+
+        limits = [
+            LimitWindowDescriptor(
+                running_total_tis_per_dagrun,
+                ["dag_id", "run_id"],
+                DagModel.max_active_tasks,
+                total_tis_per_dagrun_count,
+                TI.dag_model
+            ),
+            LimitWindowDescriptor(
+                running_tis_per_dag,
+                ["dag_id", "task_id"],
+                TI.max_active_tis_per_dag,
+                tis_per_dag_count
+            ),
+            LimitWindowDescriptor(
+                running_total_tis_per_task_run,
+                ["dag_id", "run_id", "task_id"],
+                TI.max_active_tis_per_dagrun,
+                mapped_tis_per_task_run_count,
+            ),
+            LimitWindowDescriptor(
+                running_tis_per_pool,
+                ["pool"],
+                Pool.slots,
+                pool_slots_taken,
+                TI.pool_model
+            ),
+        ]
+
+        base_query = base_query.add_columns(
+            total_tis_per_dagrun_count,
+            tis_per_dag_count,
+            mapped_tis_per_task_run_count,
+            pool_slots_taken
+        ).subquery().alias("windowed_tis")
+
+        query = (
+            select(TI)
+            .join(base_query, TI.id == base_query.c.id)
+        )
+
+        for limit in limits:
+            if limit.limit_join_model is not None:
+                query = query.join(limit.limit_join_model)
+        for limit in limits:
+            query = query.outerjoin(
+                limit.running_now_join,
+                and_(
+                    *(
+                        getattr(TI, predicate) == getattr(limit.running_now_join.c, predicate)
+                        for predicate in limit.running_now_join_predicates
+                    )
+                ),
+            )
+            query = query.where(
+                and_(
+                    func.coalesce(getattr(base_query.c, limit.window.name), text("0"))
+                    + func.coalesce(limit.running_now_join.c.now_running, text("0"))
+                    <= func.coalesce(limit.limit_column, max_tis)
+                )
+            )
+
+        query = query.options(selectinload(TI.dag_model))
+        query = query.limit(max_tis)
+        return query
```

Comment:
Have you run any performance benchmarks with this sort of query? I have to say right now I am _very_ worried about merging a change that makes _such_ a monster query just on maintainability ground, even before I've seen what it does to the DB server. I'd especially like to see what happens if you run 3 or 10 deployments against a Single DB server (in different logical databases, i.e. `CREATE DATABASE airflow_n`);

Also how does mysql cope with this query?